### PR TITLE
Init the CRDs and the controllers

### PR DESCRIPTION
**Pull request by raghavendra-talur**
This commit is the only scaffolding required for the

* 3 CRDs 
** VolumeReplicationRequest
** VolumeReplica
**  VolumeReplicationClass

* 2 controllers
** volumereplicationrequest_controller
** volumereplica_controller

### CODE REVIEW

**Comment by raghavendra-talur** [comment]
@Madhu-1 @iamniting @umangachapagain

This PR is very bare bones and the only thing to check is the domain(subdomain) names used for the resources. I would like to get this merged first so that we can discuss the CRD fields in the next PR.

File: api/v1alpha1/groupversion_info.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+package v1alpha1
+
+import (
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/scheme"
+)
+
+var (
+	// GroupVersion is group version used to register these objects
+	GroupVersion = schema.GroupVersion{Group: "ocs.openshift.io.ocs.openshift.io", Version: "v1alpha1"}
```
Comment (sp98) [inline api/v1alpha1/groupversion_info.go:29]: why is it appending the group  name twice?  (`ocs.openshift.io.ocs.openshift.io`). Is that expected ?

**Comment by Madhu-1** [comment]
@leseb @jarrpa tagging you guys to help with the review

File: api/v1alpha1/groupversion_info.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package v1alpha1 contains API Schema definitions for the ocs.openshift.io v1alpha1 API group
+// +kubebuilder:object:generate=true
+// +groupName=ocs.openshift.io.ocs.openshift.io
```
Comment (umangachapagain) [inline api/v1alpha1/groupversion_info.go:19]:
```suggestion
// +groupName=ocs.openshift.io
```

File: api/v1alpha1/volumereplica_types.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// VolumeReplicaSpec defines the desired state of VolumeReplica
+type VolumeReplicaSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// Foo is an example field of VolumeReplica. Edit VolumeReplica_types.go to remove/update
+	Foo string `json:"foo,omitempty"`
```
Comment (umangachapagain) [inline api/v1alpha1/volumereplica_types.go:32]:
```suggestion
```

File: api/v1alpha1/volumereplicationclass_types.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// VolumeReplicationClassSpec defines the desired state of VolumeReplicationClass
+type VolumeReplicationClassSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// Foo is an example field of VolumeReplicationClass. Edit VolumeReplicationClass_types.go to remove/update
+	Foo string `json:"foo,omitempty"`
```
Comment (umangachapagain) [inline api/v1alpha1/volumereplicationclass_types.go:32]:
```suggestion
```

File: api/v1alpha1/volumereplicationrequest_types.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// VolumeReplicationRequestSpec defines the desired state of VolumeReplicationRequest
+type VolumeReplicationRequestSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// Foo is an example field of VolumeReplicationRequest. Edit VolumeReplicationRequest_types.go to remove/update
+	Foo string `json:"foo,omitempty"`
```
Comment (umangachapagain) [inline api/v1alpha1/volumereplicationrequest_types.go:32]:
```suggestion
```

File: config/default/kustomization.yaml [hunk 6]
```diff
@@ -0,0 +1,74 @@
+# Adds namespace to all resources.
+namespace: volume-replication-operator-system
```
Comment (umangachapagain) [inline config/default/kustomization.yaml:2]:
```suggestion
namespace: volume-replication-operator
```

avoid using `system` suffix.

File: config/default/kustomization.yaml [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+# Value of this field is prepended to the
+# names of all resources, e.g. a deployment named
+# "wordpress" becomes "alices-wordpress".
+# Note that it should also match with the prefix (text before '-') of the namespace
+# field above.
+namePrefix: volume-replication-operator-
```
Comment (umangachapagain) [inline config/default/kustomization.yaml:9]: We need to be careful about names ending up being too long. Can we cut it short?

File: config/manager/manager.yaml [hunk 8]
```diff
@@ -0,0 +1,55 @@
+apiVersion: v1
+kind: Namespace
+metadata:
+  labels:
+    control-plane: controller-manager
```
Comment (umangachapagain) [inline config/manager/manager.yaml:5]:
```suggestion
```

File: config/manager/manager.yaml [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+  name: system
```
Comment (umangachapagain) [inline config/manager/manager.yaml:6]:
```suggestion
  name: volume-replication-operator
```

File: config/manager/manager.yaml [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: controller-manager
+  namespace: system
+  labels:
+    control-plane: controller-manager
```
Comment (umangachapagain) [inline config/manager/manager.yaml:14]:
```suggestion
```

File: config/manager/manager.yaml [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+metadata:
+  labels:
+    control-plane: controller-manager
+  name: system
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: controller-manager
+  namespace: system
```
Comment (umangachapagain) [inline config/manager/manager.yaml:12]:
```suggestion
  namespace: volume-replication-operator
```

File: config/manager/manager.yaml [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+kind: Namespace
+metadata:
+  labels:
+    control-plane: controller-manager
+  name: system
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: controller-manager
```
Comment (umangachapagain) [inline config/manager/manager.yaml:11]:
```suggestion
  name: volume-replication-operator
```

File: config/manager/manager.yaml [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+  namespace: system
+  labels:
+    control-plane: controller-manager
+spec:
+  selector:
+    matchLabels:
+      control-plane: controller-manager
```
Comment (umangachapagain) [inline config/manager/manager.yaml:18]:
```suggestion
     app: volume-replication-operator
```
probably?

File: config/manager/manager.yaml [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        control-plane: controller-manager
```
Comment (umangachapagain) [inline config/manager/manager.yaml:23]:
```suggestion
        app: volume-replication-operator
```

File: config/samples/ocs.openshift.io_v1alpha1_volumereplica.yaml [hunk 15]
```diff
@@ -0,0 +1,7 @@
+apiVersion: ocs.openshift.io.ocs.openshift.io/v1alpha1
+kind: VolumeReplica
+metadata:
+  name: volumereplica-sample
+spec:
+  # Add fields here
+  foo: bar
```
Comment (umangachapagain) [inline config/samples/ocs.openshift.io_v1alpha1_volumereplica.yaml:7]:
```suggestion
```

File: config/samples/ocs.openshift.io_v1alpha1_volumereplicationclass.yaml [hunk 16]
```diff
@@ -0,0 +1,7 @@
+apiVersion: ocs.openshift.io.ocs.openshift.io/v1alpha1
+kind: VolumeReplicationClass
+metadata:
+  name: volumereplicationclass-sample
+spec:
+  # Add fields here
+  foo: bar
```
Comment (umangachapagain) [inline config/samples/ocs.openshift.io_v1alpha1_volumereplicationclass.yaml:7]:
```suggestion
```

File: config/samples/ocs.openshift.io_v1alpha1_volumereplicationrequest.yaml [hunk 17]
```diff
@@ -0,0 +1,7 @@
+apiVersion: ocs.openshift.io.ocs.openshift.io/v1alpha1
+kind: VolumeReplicationRequest
+metadata:
+  name: volumereplicationrequest-sample
+spec:
+  # Add fields here
+  foo: bar
```
Comment (umangachapagain) [inline config/samples/ocs.openshift.io_v1alpha1_volumereplicationrequest.yaml:7]:
```suggestion
```

File: api/v1alpha1/groupversion_info.go (hunk 1 above)
Comment (raghavendra-talur) [reply api/v1alpha1/groupversion_info.go:29]: Thanks, fixed.

File: api/v1alpha1/volumereplica_types.go (hunk 3 above)
Comment (raghavendra-talur) [reply api/v1alpha1/volumereplica_types.go:32]: Would like to keep it as is in the first commit. Will remove in the next PR.

File: api/v1alpha1/volumereplicationclass_types.go (hunk 4 above)
Comment (raghavendra-talur) [reply api/v1alpha1/volumereplicationclass_types.go:32]: Would like to keep it as is in the first commit. Will remove in the next PR.

File: api/v1alpha1/volumereplicationrequest_types.go (hunk 5 above)
Comment (raghavendra-talur) [reply api/v1alpha1/volumereplicationrequest_types.go:32]: Would like to keep it as is in the first commit. Will remove in the next PR.

File: config/samples/ocs.openshift.io_v1alpha1_volumereplicationclass.yaml (hunk 16 above)
Comment (raghavendra-talur) [reply config/samples/ocs.openshift.io_v1alpha1_volumereplicationclass.yaml:7]: Would like to keep it as is in the first commit. Will remove in the next PR.

File: config/samples/ocs.openshift.io_v1alpha1_volumereplicationrequest.yaml (hunk 17 above)
Comment (raghavendra-talur) [reply config/samples/ocs.openshift.io_v1alpha1_volumereplicationrequest.yaml:7]: Would like to keep it as is in the first commit. Will remove in the next PR.

File: config/manager/manager.yaml (hunk 10 above)
Comment (raghavendra-talur) [reply config/manager/manager.yaml:14]: https://github.com/kube-storage/volume-replication-operator/issues/7

File: config/manager/manager.yaml (hunk 8 above)
Comment (raghavendra-talur) [reply config/manager/manager.yaml:5]: Created https://github.com/kube-storage/volume-replication-operator/issues/7

File: config/default/kustomization.yaml (hunk 7 above)
Comment (raghavendra-talur) [reply config/default/kustomization.yaml:9]: Created https://github.com/kube-storage/volume-replication-operator/issues/7

File: config/default/kustomization.yaml (hunk 6 above)
Comment (raghavendra-talur) [reply config/default/kustomization.yaml:2]: Created https://github.com/kube-storage/volume-replication-operator/issues/7

File: config/manager/manager.yaml (hunk 11 above)
Comment (raghavendra-talur) [reply config/manager/manager.yaml:12]: Created https://github.com/kube-storage/volume-replication-operator/issues/7

File: config/manager/manager.yaml (hunk 12 above)
Comment (raghavendra-talur) [reply config/manager/manager.yaml:11]: Created https://github.com/kube-storage/volume-replication-operator/issues/7

File: config/manager/manager.yaml (hunk 13 above)
Comment (raghavendra-talur) [reply config/manager/manager.yaml:18]: Created https://github.com/kube-storage/volume-replication-operator/issues/7

File: config/manager/manager.yaml (hunk 14 above)
Comment (raghavendra-talur) [reply config/manager/manager.yaml:23]: Created https://github.com/kube-storage/volume-replication-operator/issues/7

**Comment by raghavendra-talur** [comment]
@Madhu-1 @umangachapagain @sp98 This force push has the following changes:

1. The domain is `storage.openshift.io`. This is the closest we have in the openshift domain that is outside of the ocs group but still storage related.
2. The group is `replication`. As the domain is storage I think we don't need to state volume replication again. It is understood.
3. Now the objects are of the format kind.group.domain.
4. I have created #7 to address most of the review suggestions by @umangachapagain .

**Comment by Madhu-1** [comment]
> @Madhu-1 @umangachapagain @sp98 This force push has the following changes:
> 
> 1. The domain is `storage.openshift.io`. This is the closest we have in the openshift domain that is outside of the ocs group but still storage related.

:+1: 

> 2. The group is `replication`. As the domain is storage I think we don't need to state volume replication again. It is understood.
> 3. Now the objects are of the format kind.group.domain.
> 4. I have created #7 to address most of the review suggestions by @umangachapagain .

**Review by Madhu-1 — APPROVED** [review]
LGTM. we can do the fixes and enhancements in follow-up PR's

**Review by sp98 — APPROVED** [review]

**Review by umangachapagain — APPROVED** [review]

**Pull request merged**